### PR TITLE
test(typecheck): type-check the worker and binance __tests__ surface

### PR DESCRIPTION
**Pull request by chrisleekr**
**Stack 6/6** — base #757. Last on purpose; see below.

### The gap

`tsc -b` builds `src/` only, so **nothing ever type-checked the tests**. Fixtures drifted from the types they claimed to build and the suite stayed green.

That is not theoretical:

- Two revive tests in `apps/worker` omitted `walletQuantity` entirely, satisfied the partial at runtime, read `undefined` where the code branches on `null`, threw inside a try/catch, and **passed while asserting nothing**.
- The `SymbolInfo` fixtures had fallen behind the contract's `status`, `maxQty`, `minPrice` and `maxPrice` fields.
- Around twenty tick tests were still building `TickJobData` with `event: 'tick'`, a member `TickEvent` has not carried for some time.

### The gate

A `tsconfig.test.json` in `apps/worker` and `packages/binance` covering `src/**` plus `__tests__/**`, wired into each package's `typecheck` script. Ordinary tests and type-level tests keep separate compilers: this project owns every ordinary test, `tsconfig.test-d.json` keeps the `.test-d.ts` assertions.

### Why this is last in the stack

I originally planned it first. It cannot go there, and this was established by running it rather than by reasoning about it.

`tsconfig.test.json` includes `__tests__/**` wholesale, so it compiles **every** test file — including the ones whose fixes live in the feature commits. Arming it before them produced 100+ errors across `backfill-trade-archive`, `discovery/handler`, `reconcile-held-quantity`, `pipeline-worker`, `discovery-health.cron`, `diagnosis-worker` and `binance-rest` tests. Landing it earlier would have meant either a red gate or splitting feature test files by hunk.

### Why only these two packages

Measured, not assumed. Dropping the same 14-line project into every other workspace and compiling it surfaces **685 errors across 570 test files in 19 packages** — `apps/web` 294, `packages/strategy/trailing-trade` 137, `apps/api` 63, `packages/strategy/backtest` 38, `packages/strategy/core` 33, `packages/db` 31, `packages/contracts` 27, `packages/llm` 24, and the rest in single digits. Five are already clean and could be armed for free: `notify`, `money`, `observability`, `testcontainers`, `strategy/registry`.

`apps/server` needs a different shape — it consumes `apps/api` and `apps/worker` as project references, so a `composite: false` test project re-compiles their sources against server's `paths` and produces hundreds of spurious module-resolution errors.

### The rest

Mechanical throughout: fields added to fixtures, unions corrected, and one cast in the metrics-sink test that reproduces a JS caller now that `MetricName` is closed. **No behaviour changes.**

### Verification

Isolated worktree, fresh `bun install`: `typecheck` (with the new project confirmed to actually execute, not cache-skip), `lint`, full unit suite, and `docs:build` — 42/42 turbo tasks green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0111aiL6FyW3FtSSCjKsMzu3

## Summary by CodeRabbit

* **Tests**
  * Expanded automated coverage for account-scoped data, cache isolation, snapshot merging, order handling, event routing, and resynchronization workflows.
  * Updated test scenarios to reflect current trading, profile, notification, and strategy contracts.
  * Added stronger assertions for error handling, teardown ordering, malformed data, and persistence edge cases.

* **Chores**
  * Added comprehensive type-checking for source and test code across worker and Binance packages.
  * Improved mock and fixture consistency for more reliable test validation.

### CODE REVIEW

File: packages/binance/tsconfig.test.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    "noEmit": true,
+    "rootDir": ".",
+    "composite": false,
+    "declaration": false,
+    "declarationMap": false,
+    "isolatedDeclarations": false,
+    "incremental": false
+  },
+  "include": ["src/**/*.ts", "__tests__/**/*.ts"],
+  "exclude": ["**/*.test-d.ts"]
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply packages/binance/tsconfig.test.json]: Valid, and fixed in d2e4c5d4 by dropping the `exclude`.

Confirmed the hole is real rather than theoretical with an A/B on the same file. Dropped a deliberately broken `packages/binance/__tests__/probe.test-d.ts` (`const x: number = 'not a number'`) into the package and compiled it both ways:

- `exclude: ["**/*.test-d.ts"]` present → exit 0, silent.
- `exclude` dropped → `__tests__/probe.test-d.ts(1,7): error TS2322: Type 'string' is not assignable to type 'number'.`

Same file, same compiler, opposite outcomes. It fails open exactly as you describe.

One correction on the alternative: I tried adding the companion `tsconfig.test-d.json` and it does **not** error on empty input the way I expected — its `include` also carries `src/**/*.ts`, so the project has inputs and exits 0. It works, it is just a second full compile of `src` to cover zero files. Dropping the `exclude` gets the same coverage for free, so that is what landed.

Dropping it has a second effect worth noting: `exclude` was **overriding** the base config rather than extending it, so `["dist", "node_modules"]` from `packages/binance/tsconfig.json` was being discarded. Inert today because `include` only globs `src/**` and `__tests__/**`, but it is gone now. `apps/worker` still overrides, and there the exclusion is load-bearing, so it stays.

Added a short comment recording why this package carves out nothing while `apps/worker` does.

File: apps/worker/tsconfig.test.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  "compilerOptions": {
+    "noEmit": true,
+    "rootDir": ".",
+    "composite": false,
+    "declaration": false,
+    "declarationMap": false,
+    "isolatedDeclarations": false,
+    "incremental": false
+  },
+  "include": ["src/**/*.ts", "__tests__/**/*.ts"],
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply apps/worker/tsconfig.test.json:15]: Right that the title overreached; the title is now `test(typecheck): type-check the worker and binance __tests__ surface`, and the body carries the measurement below.

Your count is low, though. I measured it rather than estimating: dropped the same 14-line project into every other workspace and compiled it.

| package | test files | errors |
| --- | --- | --- |
| `apps/web` | 196 | 294 |
| `packages/strategy/trailing-trade` | 63 | 137 |
| `apps/api` | 78 | 63 |
| `packages/strategy/backtest` | 22 | 38 |
| `packages/strategy/core` | 19 | 33 |
| `packages/db` | 72 | 31 |
| `packages/contracts` | 32 | 27 |
| `packages/llm` | 5 | 24 |
| `packages/indicators` | 16 | 7 |
| `packages/core` | 14 | 5 |
| `packages/strategy/momentum` | 10 | 16 |
| `packages/strategy/rebalance` | 8 | 8 |
| `packages/discovery` | 9 | 2 |

**570 uncovered test files, 685 errors, 19 packages** — not ~204 files. Nearly 3x, which matters for how the follow-up gets sized: this is not "extend the same 14-line config outward", it is a real remediation project best done one package per PR.

Two things the sweep turned up that are worth having on the record:

**Five packages are already clean** and could be armed for zero fixing: `notify`, `money`, `observability`, `testcontainers`, `strategy/registry`. Small (17 files between them), but it locks the property in before it can rot. I left them out of this PR to keep it to what the review asked for.

**`apps/server` needs a different shape.** My probe reported 693 errors there, and that number is an artifact — I am flagging it so nobody chases it. It consumes `apps/api` and `apps/worker` as project references, and `composite: false` discards those, so `tsc` recompiles their sources against server's own `paths` and produces hundreds of spurious `TS2307`s. Its `tsconfig.json` comment already warns about exactly this. A test project there has to preserve `composite`.

Not opening the follow-up issue myself, since that is the repo owner's call.

File: apps/worker/__tests__/profile-manager/profile-manager.test.ts [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
 const u = asUserId;
 const p = asProfileId;
+const a = asAccountId;
+
+type TestProfileRow = Omit<ProfileLoadRow, 'operatorId'>;
+
+const scoped = (row: TestProfileRow): ProfileLoadRow => ({
+  ...row,
+  operatorId: row.userId,
+});
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply apps/worker/__tests__/profile-manager/profile-manager.test.ts]: Correct, and this was the best catch of the round. Fixed in d2e4c5d4.

You are right that it is a real gap, and I proved it before changing anything. `profile-manager.ts:199` copies `operatorId: row.operatorId` into the active-profile map, and it surfaces through `listActive()` and `operatorOf()` into the cron producers that build `scopeProfile`. So I mutated production to mis-route it:

```
-      operatorId: row.operatorId,
+      operatorId: row.userId,
```

**27 passed (27).** Silent. A PR whose entire thesis is "a fixture that drifts from its type passes for the wrong reason" had shipped exactly that, spelled `Omit` instead of `Partial`.

The `listActive` expectation did already assert `operatorId`, which is what makes it interesting — it asserted `operatorId: 'u1'`, and that was true only because `scoped` forced it. The assertion existed and was vacuous.

Fixed as you suggested: `TestProfileRow` and `scoped` are gone, `makePm` takes `readonly ProfileLoadRow[]`, and all 24 fixture rows carry an `operatorId` distinct from their `userId` (`u1` → `o1`, `u2` → `o2`), with the `listActive` expectations following. Same mutation now:

**1 failed | 26 passed**, on `"operatorId": "o1"` vs `"u1"`.

One refinement to the reasoning. Your comment says this "keeps the test compiler able to detect missing or incorrectly routed operator identity" — the compiler catches **missing** (that is what the `Omit` was suppressing), but it can never catch **mis-routed**, because `userId` and `operatorId` are both branded `UserId` and therefore mutually assignable. Only a distinct fixture value plus an assertion closes that half, which is why the fix needed the distinct values and not just the type change.

Worth noting the new gate caught my own incomplete edit here: five further `scoped(...)` call sites I had missed failed `tsc -p tsconfig.test.json` with `TS2304: Cannot find name 'scoped'` before any test ran.

File: apps/worker/__tests__/queues/backtest-worker.test.ts [hunk 4]
```diff
@@ -2,15 +2,24 @@ import { describe, expect, it, vi, beforeEach } from 'vitest';
...
-  updateProgress: vi.fn(async () => undefined),
-  complete: vi.fn(async () => undefined),
-  fail: vi.fn(async () => undefined),
-  get: vi.fn(async () => null as { status: string; symbols?: readonly string[] } | null),
+  markRunning: vi.fn<BoundMethod<'markRunning', boolean>>(async () => true),
+  updateProgress: vi.fn<BoundMethod<'updateProgress', void>>(async () => undefined),
+  complete: vi.fn<BoundMethod<'complete', void>>(async () => undefined),
+  fail: vi.fn<BoundMethod<'fail', void>>(async () => undefined),
+  get: vi.fn<BoundMethod<'get', { status: string; symbols?: readonly string[] } | null>>(
+    async () => null,
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply apps/worker/__tests__/queues/backtest-worker.test.ts]: Both valid. Fixed in d2e4c5d4.

**Repo mocks.** `BoundMethod` is gone; all five derive from `ProfileRepo['backtestRuns'][K]` whole. Checking the returns it was hard-coding: `markRunning` → `Promise<boolean>`, `updateProgress`/`complete`/`fail` → `Promise<void>`. Four of the five overrides were restating the real type, so deriving them costs nothing.

`get` was the one real divergence, and not accidental — it narrowed `Promise<BacktestRunRow | null>` to `{ status: string; symbols?: readonly string[] } | null` so the four `mockResolvedValue` sites could pass a two-field stand-in. Deriving it produced exactly that:

```
error TS2740: Type '{ status: string; symbols: string[]; }' is missing the following
properties from type 'BacktestRunRow': backtestSignature, configFingerprint,
createdAt, error, and 9 more.
```

So it now builds a real row through a `runRow(over?: Partial<BacktestRunRow>)` fixture. That is the point of the PR — a column the worker reads can no longer vanish from the row type without failing here.

**`emitEvent`.** Derived from `typeof import('../../src/executor/event-emitter.js').emitEvent`. Verified the branding actually binds rather than assuming the generic survived `vi.fn`: a probe reading `mock.calls[0]![3]` into a bogus literal type reports

```
Type '"backtest-complete" | "backtest-progress" | "heartbeat" | "logs" | "orders"
| "profile-state" | "resync-required" | "symbol-state"' is not assignable to type
'"definitely-not-a-ws-topic"'.
```

`c[3]` is the closed `WsTopic` union now, not `string`, so the `completeTopics` and `backtest-progress` assertions are checked against the real topic set. `AccountId`/`ProfileId` are branded again too.

12/12 green, and `tsc -p tsconfig.test.json` clean.

File: apps/worker/__tests__/crons/discovery/run.test.ts [hunk 5]
```diff
@@ -135,10 +135,21 @@ const fakePort = (over: Partial<DiscoveryProfilePort> = {}): DiscoveryProfilePor
...
+const snapshotFor = (
+  port: DiscoveryProfilePort,
+): Parameters<DiscoveryProfilePort['persistSnapshot']>[0] => {
+  const persistSnapshot = port.persistSnapshot as ReturnType<
+    typeof vi.fn<DiscoveryProfilePort['persistSnapshot']>
+  >;
+  const snapshot = persistSnapshot.mock.calls[0]?.[0];
+  if (!snapshot) throw new Error('expected a persisted discovery snapshot');
+  return snapshot;
+};
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply apps/worker/__tests__/crons/discovery/run.test.ts:152]: Taking the useful half of this, declining the tag scaffolding. d2e4c5d4.

**Added** a one-line WHY to the two helpers here that are genuinely non-obvious, because in both cases the *type* is the puzzle:

- `snapshotFor` — why the cast exists at all: `DiscoveryProfilePort` says nothing about mocks, so reading `.mock.calls` needs the cast back to the spy type the port was built with.
- `fetchSpy` — why `Object.assign`: `fetchImpl` is typed as the global `fetch`, which carries a `preconnect` property a bare `vi.fn` lacks, so the spy grafts the real one on.

That is the clause of gate #9 that earns its keep here: *"The prose above the tags carries the WHY."*

**Declining** the `@param`/`@returns` blocks. On these three the tags would read `@param port - The port`, `@param implementation - The implementation`, `@param input - The input` — which is the *"no name-restatements (`@param symbol - The symbol`)"* the same gate explicitly bans. Adding them would violate the rule they are cited to satisfy.

`makePm` I have left alone deliberately: it already carries a three-line comment explaining the boot back-edge wire it mirrors, which is the part a reader cannot infer. It also changed substantially in d2e4c5d4 for the `operatorId` finding.

It is also not the convention here — 71 of 225 worker test files contain any `/**` block at all, against 458 arrow-function helpers. Applying a `src` rule to local fixture builders across the suite is a repo-wide decision, not something to settle inside a test-typing PR.

File: apps/worker/__tests__/crons/orphan-orders-detect.test.ts [hunk 6]
```diff
@@ -28,18 +29,18 @@ const job = { id: 'job-1', data: {} } as unknown as Job;
...
+
+const profile = (profileId: string, accountId = ACC_LIVE): ActiveProfile => ({
+  profileId: asProfileId(profileId),
+  accountId,
+  userId: asUserId('u1'),
+  operatorId: asUserId('u1'),
+  candleInterval: '1h',
+  symbols: ['BTCUSDT'],
+  technicalsIntervals: [],
+});
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply apps/worker/__tests__/crons/orphan-orders-detect.test.ts:43]: Declining this one. Reasoning as on the sibling thread, plus a scope point that is decisive here.

Six of the eight helpers listed are **not touched by this PR**. Checked each against `origin/main`:

| helper | added by this PR | already on `main` |
| --- | --- | --- |
| `notifyAll` | no | yes |
| `buildExecutor` | no | yes |
| `buildStubStrategy` | no | yes |
| `buildColdLoad` | no | yes |
| `buildMarketDataPort` | no | yes |
| `stratWith` | no | yes |
| `profile` | modified | yes |
| `firstOrder` | yes | no |

The diff on the six is fixture fields added *inside* the call sites — the helpers themselves are untouched. Documenting pre-existing code because a nearby line moved is unrelated scope in a PR that states no behaviour changes, and gate #9 takes the same line for its wrapping clause: *"existing wrapped comments are left alone until the code around them is touched."*

On the two this PR does own: `firstOrder(input: OpenOrderDto)` returns the first mapped order and throws if there is none — `@param input - the DTO` / `@returns the order` adds nothing a reader does not already have from the signature, and is the name-restatement the gate bans. `profile` gained an `accountId` field; its shape is self-describing.

I did add WHY comments to the two new helpers on the sibling thread where the *type* is genuinely puzzling (`snapshotFor`'s mock cast, `fetchSpy`'s `preconnect` graft). That is the part of this finding with real value, and it is in d2e4c5d4.

If you want fixture-builder JSDoc to become the convention, that is a repo-wide sweep across 458 helpers in `apps/worker` alone — worth its own PR, not this one.

File: apps/worker/__tests__/profile-manager/profile-manager.test.ts [hunk 7]
```diff
@@ -37,7 +39,9 @@ describe('ProfileManager', () => {
   // Mirror the boot back-edge wire: construct the manager, then inject the
   // market back-edge before start(). The account user-data stream is driven by
   // subscription-ownership, not the manager, so there is no stream back-edge.
-  const makePm = (deps: Parameters<typeof createProfileManager>[0]): ProfileManager => {
+  const makePm = (deps: {
+    loadEnabledProfiles: () => Promise<readonly ProfileLoadRow[]>;
+  }): ProfileManager => {
     const pm = createProfileManager(deps);
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply apps/worker/__tests__/profile-manager/profile-manager.test.ts:45]: Declining, same reasoning as the two sibling JSDoc threads. Two of the three already carry the part of gate #9 that matters:

**`makePm`** has a three-line block directly above it, unchanged from `main`:

> Mirror the boot back-edge wire: construct the manager, then inject the market back-edge before `start()`. The account user-data stream is driven by subscription-ownership, not the manager, so there is no stream back-edge.

That is the WHY, and it is the only thing about this helper a reader cannot get from the signature. `@param deps - the deps` / `@returns a ProfileManager` would sit above it adding nothing.

**`runRow`** carries a WHY I added in d2e4c5d4:

> `get` returns a full `BacktestRunRow`, so the mock builds one rather than a two-field shape: narrowing the mock's return would let a column the worker reads disappear from the row type without failing here.

That sentence is the entire reason the helper exists — it is the fix for your own `backtest-worker` finding. `@param over - partial overrides` is strictly less informative.

**`row`** is an eight-line object literal returning `ProfileLoadRow`, with `profileId`, `symbols` and `candleInterval` named in the signature and echoed verbatim in the body.

Gate #9 asks for a block whose prose carries the WHY and whose tags gloss what each value *means*, and in the same breath bans name-restatements. For these three the tags have nothing to gloss — every parameter is a typed pass-through to an identically-named field. Adding them would satisfy the letter of the rule by violating the clause that gives it its purpose.

As noted on the sibling threads: 71 of 225 worker test files contain any `/**` at all, against 458 arrow-function helpers. Making fixture-builder JSDoc the convention is a repo-wide call, not one to settle inside a test-typing PR.
